test/ssl/client_ssl_authentication_test.go

```diff
@@ -0,0 +1,226 @@
+		t.Error(err)
+	}
+}
+
+func TestSSLMutualAuthentication_ServerDoesntKnowClientFail(t *testing.T) {
+	clusterID, err := createMemberWithXML(maRequiredXML)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer remoteController.ShutdownCluster(clusterID)
+	config, err := createClientConfigWithSSLConfig(client2Cert,
+		client2Key, server2CA)
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = hazelcast.NewClientWithConfig(config)
+	if _, ok := err.(*core.HazelcastIllegalStateError); !ok {
+		t.Error(err)
+	}
+}
+
+func TestSSLMutualAuthentication_NeitherServerNorClientKnowsTheOtherFail(t *testing.T) {
+	clusterID, err := createMemberWithXML(maRequiredXML)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer remoteController.ShutdownCluster(clusterID)
+	config, err := createClientConfigWithSSLConfig("client2-cert.pem",
+		client2Key, server2CA)
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = hazelcast.NewClientWithConfig(config)
+	if _, ok := err.(*core.HazelcastIllegalStateError); !ok {
+		t.Error(err)
+	}
+}
+
+func TestSSLOptionalMutualAuthenticationConnect(t *testing.T) {
+	clusterID, err := createMemberWithXML(maOptionalXML)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer remoteController.ShutdownCluster(clusterID)
+	config, err := createClientConfigWithSSLConfig(client1Cert,
+		client1Key, server1CA)
+	if err != nil {
+		t.Fatal(err)
+	}
+	client, err := hazelcast.NewClientWithConfig(config)
+	defer client.Shutdown()
+
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestSSLOptionalMutualAuthentication_ClientDoesntKnowServerFail(t *testing.T) {
+	clusterID, err := createMemberWithXML(maOptionalXML)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer remoteController.ShutdownCluster(clusterID)
+	config, err := createClientConfigWithSSLConfig(client1Cert,
+		client1Key, server2CA)
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = hazelcast.NewClientWithConfig(config)
+	if _, ok := err.(*core.HazelcastIllegalStateError); !ok {
+		t.Error(err)
+	}
+}
+
+func TestSSLOptionalMutualAuthentication_ServerDoesntKnowClientConnect(t *testing.T) {
+	clusterID, err := createMemberWithXML(maOptionalXML)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer remoteController.ShutdownCluster(clusterID)
+	config, err := createClientConfigWithSSLConfig(client2Cert,
+		client2Key, server1CA)
+	if err != nil {
+		t.Fatal(err)
+	}
+	config.NetworkConfig().SSLConfig().Certificates = nil
+	client, err := hazelcast.NewClientWithConfig(config)
+	defer client.Shutdown()
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestSSLOptionalMutualAuthentication_NeitherServerNorClientKnowsTheOther(t *testing.T) {
+	clusterID, err := createMemberWithXML(maOptionalXML)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer remoteController.ShutdownCluster(clusterID)
+	config, err := createClientConfigWithSSLConfig(client2Cert,
+		client2Key, server2CA)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	_, err = hazelcast.NewClientWithConfig(config)
+	if _, ok := err.(*core.HazelcastIllegalStateError); !ok {
+		t.Error(err)
+	}
+}
```

Comment:
there should be a test where you set mutual authentication to optional and provide no client certificates